src/nccl_ofi_net.c

```diff
@@ -158,85 +158,204 @@ void nccl_net_ofi_free_info_list(struct fi_info *info_list)
 }
 
 /*
- * @brief	Allocate a memory registration key
+ * @brief	Initialize pool of IDs
  *
- * Extract an available key from the key pool, mark the key as
- * unavailable in the key pool, and return extracted key. Noop in case
- * no key was available.
+ * Initializes a nccl_ofi_id_pool_t object, marking all IDs as available.
  *
- * This operation is locked by the key pool's internal lock.
+ * @param	id_pool
+ *		The ID pool
+ * @param	size
+ *		Size of the id pool (number of IDs)
+ * @return	0 on success
+ *		non-zero on error
+ */
+int nccl_ofi_id_pool_init(nccl_ofi_id_pool_t *id_pool, size_t size)
+{
+	int ret = 0;
+
+	if (OFI_UNLIKELY(NULL == id_pool)) {
+		NCCL_OFI_WARN("Invalid call to nccl_ofi_id_pool_init with null pointer");
+		return -EINVAL;
+	}
+
+	/* Scale pool size to number of 64-bit entries (rounded up) */
+	id_pool->entries = (size + sizeof(uint64_t) * 8 - 1) / (sizeof(uint64_t) * 8);
+
+	/* Allocate memory for the pool */
+	id_pool->ids = malloc(sizeof(uint64_t) * id_pool->entries);
+
+	/* Return in case of allocation error */
+	if (OFI_UNLIKELY(NULL == id_pool->ids)) {
+		NCCL_OFI_WARN("Unable to allocate ID pool");
+		return -ENOMEM;
+	}
+
+	/* Set all IDs to be available */
+	memset(id_pool->ids, 0xff, size / 8);
+	if (size % 8) {
+		id_pool->ids[id_pool->entries - 1] = (1ULL << (size % (sizeof(uint64_t) * 8))) - 1;
```

Comment:
Nit: If you were tracking 0 as available instead, a calloc would have given you malloc+memset.

Reply:
that is true, but the reason I had to make available =1 is because I wanted to use a find_first_set instruction to get an available ID from the bit array. With available =0 I would need to negate the array before using FFS, but this looked like a cleaner way (initialize to one once at the beginning vs do a negation every time we allocate it).